test/quadiron_c_utest.cpp

```diff
@@ -0,0 +1,246 @@
+        std::vector<int> selector(n);
+        std::fill(selector.begin(), selector.begin() + k, 1);
+        do {
+            for (int i = 0; i < n; i++) {
+                if (selector[i]) {
+                    selected.push_back(i);
+                }
+            }
+            combinations.push_back(selected);
+            selected.clear();
+        } while (std::prev_permutation(selector.begin(), selector.end()));
+        return combinations;
+    }
+
+    /** Convert list of indexes to list of boolean values
+     *
+     * @param src input idx_list
+     * @param n_src number of entries in src
+     * @param dst output idx list as boolean values (1/0)
+     * @param n_dst number of entries in dst
+     */
+    void convert_idx_list(
+        std::vector<int> src,
+        int n_src,
+        std::vector<int>& dst,
+        int n_dst)
+    {
+        std::fill_n(dst.begin(), n_dst, 0);
+        for (int i = 0; i < n_src; i++) {
+            ASSERT_LT(src[i], n_dst);
+            dst[src[i]] = 1;
+        }
+    }
+
+    /** Test encode/decode/reconstruct
+     *
+     * This test will create n_data fragments and generate n_parities parities,
+     * then decode with a set of missing fragments. Finaly reconstruct them.
+     *
+     * @param n_data number of data
+     * @param n_parities number of parities
+     * @param block_size size of block in bytes
+     * @param systematic 1 if systematic else 0
+     * @param missing_idxs vector of boolean vales indicating missing fragments
+     * for decode and reconstruct
+     *        - must be of length n_parities
+     */
+    void test_encode_decode_reconstruct(
+        int n_data,
+        int n_parities,
+        size_t block_size,
+        int systematic,
+        std::vector<int> missing_idxs)
+    {
+        struct QuadironFnt32* inst =
+            quadiron_fnt32_new(2, n_data, n_parities, systematic);
+        size_t metadata_size =
+            quadiron_fnt32_get_metadata_size(inst, block_size);
+        std::vector<std::vector<uint8_t>> data(n_data);
+        std::vector<uint8_t*> _data(n_data); // for C API
+        std::vector<std::vector<uint8_t>> ref_data(n_data);
+        std::vector<uint8_t*> _ref_data(n_data); // for C API
+        int n_outputs;
+        if (systematic) {
+            n_outputs = n_parities;
+        } else {
+            n_outputs = n_data + n_parities;
+        }
+        std::vector<std::vector<uint8_t>> parity(n_outputs);
+        std::vector<uint8_t*> _parity(n_outputs); // for C API
+        std::vector<int> wanted_idxs(n_outputs);
+
+        for (int i = 0; i < n_data; i++) {
+            data.at(i).resize(block_size + metadata_size);
+            _data[i] = data.at(i).data();
+            ref_data.at(i).resize(block_size);
+            _ref_data[i] = ref_data.at(i).data();
+            randomize_buffer(_data[i] + metadata_size, block_size);
+            std::memcpy(_ref_data[i], _data[i] + metadata_size, block_size);
```

Comment:
should be `std::copy_n(_data[i] + metadata_size, block_size, _ref_data[i]);` (tested locally).